doc/manual/contract.md

```diff
@@ -0,0 +1,505 @@
+
+The implementation of contracts is recommended when we are dealing with business rules. For example when we are developing a library it may be necessary to specify contracts to guarantee valid usages of wer API. Also, contracts support inheritance, which guarantees that contracts will be satisfied in future specification.
+
+Let us take a look at the `Stack` class bellow:
+
+~~~
+class Stack[T]
+	invariant(size >= 0)
+
+	var data = new Array[T]
+
+	var size = 0
+
+	fun is_empty: Bool
+	is
+		ensure(result == (size == 0))
+	do
+		return size == 0
+	end
+
+	# Add `item` on top of the stack
+	fun push(item: T)
+	is
+		ensure(size == old(size) + 1)
+		ensure(data[size-1] == item)
+	do
+		data[size] = item
+		size += 1
+	end
+
+	# Remove the item on top of the stack
+	fun pop
+	is
+		expect(not is_empty)
+		ensure(size == old(size) - 1)
+	do
+		data.pop
+		size -= 1
+	end
+
+	# Return the item at the given index.
+	fun [](index: Int): T
+	is
+		expect(index < size)
+		expect(index >= 0)
+		ensure(result == data[index])
+	do
+		return data[index]
+	end
+
+	# Set the item at the given index
+	fun []=(index: Int, item: T)
+	is
+		expect(index >= 0)
+		expect(index <= size)
+	do
+		data[index] = item
+	end
+end
+~~~
+
+As we can see the contracts are defined using several annotations: `invariant` which designates the class invariants, `expect` which designates the preconditions and `ensure` which designates the postconditions. The conditions of a contract must be expressed using boolean expressions in the same way as assertions. Each contract of this example will be detailed in the sections below.
+
+By default when you compile or interpret (`nitc stack.nit` or `nit stack.nit`) a program, all contracts (`ensure`, `expect`, `invariant`) used in the main module are enabled, direct imported modules have only the expects contracts enabled (`ensure` and `invariant` contracts are disabled). The other indirect imported modules don't have active contracts. We can use options to change the contract rule see `Options` section for more information.
+
+All contracts conditions can be declared in several ways, all are equivalent:
+~~~
+fun foo(a: Int, b: Int) is expect(a > 0, b < 0)
+~~~
+~~~
+fun foo(a: Int, b: Int) is expect(a > 0 and b < 0)
+~~~
+~~~
+fun foo(a: Int, b: Int) is expect(a > 0), ensure(b < 0)
+~~~
+
+## Writing contracts
+
+Methods, as well as attribute setters, can be associated with two types of contracts; precondition (`expect`) that expresses requirements of the routine´s call and postcondition (`ensure`) that expresses what the routine guarantee after her execution. Together, they define the rules to be respected between the callee and the caller. If the caller fulfills the `expect` condition on entry then the caller can assume that the callee fulfill the `ensure` condition on exit.
```

Comment:
Didn't I read the same paragraph but worded differently line 68? I prefer this version tho.

Reply:
The line 68 only introduces the annotations with their correspondence (`expect` = precondition, `ensure` = postcondition, `invariant` = invariant). Here I add more explanations. Do you think it's redundant?